components/server/src/bitbucket/bitbucket-repository-provider.ts

```diff
@@ -59,24 +59,40 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
     async getBranches(user: User, owner: string, repo: string): Promise<Branch[]> {
         const branches: Branch[] = [];
         const api = await this.apiFactory.create(user);
-        const response = await api.repositories.listBranches({
-            workspace: owner,
-            repo_slug: repo,
-            sort: "target.date",
-        });
 
-        for (const branch of response.data.values!) {
-            branches.push({
-                htmlUrl: branch.links?.html?.href!,
-                name: branch.name!,
-                commit: {
-                    sha: branch.target?.hash!,
-                    author: branch.target?.author?.user?.display_name!,
-                    authorAvatarUrl: branch.target?.author?.user?.links?.avatar?.href,
-                    authorDate: branch.target?.date!,
-                    commitMessage: branch.target?.message || "missing commit message",
-                },
+        // Handle pagination.
+        let nextPage = 1;
+        let isMoreDataAvailable = true;
+
+        while (isMoreDataAvailable) {
```

Comment:
Are there risk that hit some rate limitting here?

Reply:
Good question! 🎯 Probably yes (just like all other Bitbucket API requests), but I think using a `pagelen` of `100` mitigates this somewhat (i.e. you'd have to have an incredibly large number of branches in order to get into rate limit territory, and we also do all these requests in sequence and not in parallel).

My feeling here is that it's safe to let any (supposedly very rare & unlikely) rate limit error just bubble up as is if it ever happens, i.e. the Gitpod user would see something like "Unable to fetch branches: Bitbucket API rate limit exceeded".

Reply:
For repositories we have pagination as well but it broke with 10000 as far as I understand and was very slow to fetch everything breaking ability to create a project.

Reply:
@akosyakov aha -- was that with `pagelen: 100` or the default page length?

Reply:
I think it was 10K repos with 1000 page size